pkg/reconciler/identity/iam/iam_policy.go

```diff
@@ -0,0 +1,276 @@
+		role:           iam.RoleName(role),
+		member:         member,
+		action:         actionRemove,
+		respCh:         make(chan error, 1),
+	})
+}
+
+func (m *manager) doRequest(ctx context.Context, req *modificationRequest) error {
+	select {
+	case m.requestCh <- req:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	select {
+	case err := <-req.respCh:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// manage receives requests on m.requestCh and adds their modifications to the service account's
+// modification batch in m.pending. When a new batch is created, manage will initiate a call to
+// GetIAMPolicy which will return its result on m.getPolicyCh. When manage receives a policy on
+// getPolicyCh it will apply all batched modifications to that policy and initiate a call to
+// SetIAMPolicy which will also return its result m.getPolicyCh. When there are no batched
+// modifications to apply to a policy, manage will instead discard the policy and delete the service
+// account's entry in m.pending.
+func (m *manager) manage(ctx context.Context) {
+	for {
+		select {
+		case req := <-m.requestCh:
+			if err := m.makeModificationRequest(ctx, req); err != nil {
+				req.respCh <- err
+			}
+		case getPolicy := <-m.getPolicyCh:
+			batched := m.pending[getPolicy.account]
+			if len(batched.listeners) == 0 {
+				delete(m.pending, getPolicy.account)
+				break
+			}
+			if getPolicy.err != nil {
+				for _, listener := range batched.listeners {
+					listener <- getPolicy.err
+				}
+				delete(m.pending, getPolicy.account)
+				break
+			}
+			m.pending[getPolicy.account] = &batchedModifications{
+				roleModifications: make(map[iam.RoleName]*roleModification),
+			}
+			go m.applyBatchedModifications(ctx, getPolicy.account, getPolicy.policy, batched)
+		case <-ctx.Done():
+			for _, batched := range m.pending {
+				for _, listener := range batched.listeners {
+					listener <- ctx.Err()
+				}
+			}
+			return
+		}
+	}
+}
+
+// makeModificationRequest adds the modification request to the service account's existing batch if
+// one exists. Otherwise it will create a new batch and start a call to getPolicy.
+func (m *manager) makeModificationRequest(ctx context.Context, req *modificationRequest) error {
+	batched := m.pending[req.serviceAccount]
+	var mod *roleModification
+	if batched != nil {
+		mod = batched.roleModifications[req.role]
+	}
+	if mod == nil {
+		mod = &roleModification{
+			addMembers:    sets.NewString(),
+			removeMembers: sets.NewString(),
+		}
+		if batched != nil {
+			batched.roleModifications[req.role] = mod
+		}
+	}
+	switch req.action {
+	case actionAdd:
+		if mod.removeMembers.Has(req.member) {
+			return fmt.Errorf("conflicting remove of member %s", req.member)
+		}
+		mod.addMembers.Insert(req.member)
+	case actionRemove:
+		if mod.addMembers.Has(req.member) {
+			return fmt.Errorf("conflicting add of member %s", req.member)
+		}
+		mod.removeMembers.Insert(req.member)
+	}
+	if batched == nil {
```

Comment:
It seems like if you move this block to the beginning, the logic is more clear and can hopefully simplify the `if mod==nil` and `if bached !=nil` checks. I understand that you may not need to run this if there is an error.

Reply:
Done.